feat: add password-protected file uploads

anoldguy opened the pull request:
## Description

This PR adds password-protected file uploads to rustypaste. Users upload files with the `protected` form field, the server generates a random password and returns it alongside the URL, and downloads require that password via Authorization header.

I tried to follow the existing paste type patterns closely. Passwords are hashed with Argon2id and stored in sidecar files (`filename.txt.password`) alongside the uploaded content. This avoids extended attributes or database dependencies while keeping cleanup straightforward.

File operations use careful ordering to prevent unprotected content exposure. Creation stores password hashes before writing content files, and reverse that order for deletion. If either operation is interrupted, worst case is an orphaned password file with no exposed content. The cleanup routine removes these orphaned files when detected.

## Motivation and Context

Sharing sensitive files requires more protection than obscurity through random URLs. Password protection adds a basic authentication layer while keeping rustypaste's simplicity intact.

This implementation aims to fit naturally into the existing codebase rather than introduce new architectural patterns. The sidecar file approach mirrors how expiration timestamps are handled and makes cleanup deterministic.

## Design Decisions

**Storage location:** Protected files are stored in a dedicated `protected/` subdirectory, following the same pattern as oneshot and URL paste types. This provides natural isolation and simplifies the codebase by eliminating the need for special filtering logic in the list endpoint (my first approach).

## How Has This Been Tested?

### Manual Testing

Tested locally against the development server:

- Upload with `curl -F "protected=@file.txt"` returns URL and password
- Download with Bearer token succeeds: `curl -H "Authorization: Bearer <password>"`
- Download with Basic Auth succeeds: `curl -u "user:<password>"`
- All failure modes return 404 to prevent enumeration (missing auth, wrong password, or non-existent file all return identical responses)
- Cleanup removes both main file and password sidecar
- Orphaned password files are identified and removed

### Automated Tests

**Unit tests** (`src/password.rs`):
- `test_password_hashing` - Verifies hash generation and password verification against Argon2
- `test_password_file_path` - Tests path transformation logic for password sidecar files
- `test_password_file_path_invalid_utf8` - Validates error handling for paths with invalid UTF-8 sequences
- `test_store_and_verify_password` - Full roundtrip test covering storage, verification, and cleanup

**Integration tests** (`src/server.rs`):
- `test_upload_protected_file` - Tests upload/download flow with all authentication scenarios (missing auth, wrong password via Bearer/Basic, correct password via Bearer/Basic)
- `test_protected_file_enumeration_prevention` - Verifies all authentication failures return identical 404 responses to prevent file enumeration
- `test_list_excludes_password_files` - Confirms protected files and their password sidecars are hidden from the list endpoint

Tests cover the critical path and error handling for password-protected files. All existing tests pass.

## Changelog Entry

### Added

- Password-protected file uploads via `protected` form field
  - Server generates 24-character alphanumeric passwords
  - Passwords hashed with Argon2id (19MB memory, 2 iterations)
  - Download requires password via Authorization header (Bearer or Basic Auth)
  - All authentication failures return 404 to prevent file enumeration
  - Protected files hidden from list endpoint to prevent enumeration
  - Cleanup handles orphaned password files

## Types of Changes

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the code style of this project.
- [x] I have updated the documentation accordingly.
- [x] I have formatted the code with [rustfmt](https://github.com/rust-lang/rustfmt).
- [x] I checked the lints with [clippy](https://github.com/rust-lang/rust-clippy).
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

---

**Note to maintainer:** I'm still building familiarity with Rust, so if there are patterns here that could be more idiomatic or align better with the project's conventions, I'm happy to revise. Figuring out the tests needed to be run single-threaded was an interesting experience. I thought I had broken everything before I dug into the `ci` workflow to see how they ran 'for real'.

anoldguy committed:
feat: add password-protected file uploads

Added a ProtectedFile paste type that auto-generates passwords on upload. I'm still getting comfortable with Rust idioms, so I followed the existing paste type patterns closely. Password hashes live in sidecar files (filename.txt.password) alongside the uploaded content, avoiding the complexity of xattrs or databases.

Upload with the "protected" form field:

    curl -F "protected=@secret.txt" http://localhost:8080

The server generates a random 24-character password and returns both URL and password:

    http://localhost:8080/secret.txt
    Password: aBcD1234EfGh5678IjKl9012

Downloads require the password via Authorization header (Bearer or Basic Auth):

    curl -H "Authorization: Bearer aBcD1234EfGh5678IjKl9012" http://localhost:8080/secret.txt
    curl -u "user:aBcD1234EfGh5678IjKl9012" http://localhost:8080/secret.txt

Security: Argon2id hashing (19MB memory, 2 iterations). Wrong passwords return 404 instead of 401 to prevent file enumeration. Cleanup routines handle orphaned password files when main files are deleted or expire.

Tests pass for password generation, hashing, verification, and file management.

anoldguy committed:
feat: add password-protected file uploads

Added a ProtectedFile paste type that auto-generates passwords on upload. I'm still getting comfortable with Rust idioms, so I followed the existing paste type patterns closely. Password hashes live in sidecar files (filename.txt.password) alongside the uploaded content, avoiding the complexity of xattrs or databases.

Upload with the "protected" form field:

    curl -F "protected=@secret.txt" http://localhost:8080

The server generates a random 24-character password and returns both URL and password:

    http://localhost:8080/secret.txt
    Password: aBcD1234EfGh5678IjKl9012

Downloads require the password via Authorization header (Bearer or Basic Auth):

    curl -H "Authorization: Bearer aBcD1234EfGh5678IjKl9012" http://localhost:8080/secret.txt
    curl -u "user:aBcD1234EfGh5678IjKl9012" http://localhost:8080/secret.txt

Security: Argon2id hashing (19MB memory, 2 iterations). Wrong passwords return 404 instead of 401 to prevent file enumeration. Cleanup routines handle orphaned password files when main files are deleted or expire.

Tests pass for password generation, hashing, verification, and file management.

tessus reviewed (commented):
Nice PR. The code is pretty straight forward and also thank you for the extensive tests.

I have 3 comments:

- I am not the owner of this project and it is not my decision whether this is merged or not. But I would, if it were.
- please add a [fixture](https://github.com/orhun/rustypaste/tree/master/fixtures) test
- support for protected files in the cli app would also be nice (not a requirement for this PR)

Let's wait for @orhun's input.

anoldguy commented:
1. Smart idea about a fixture test. I'll get that wired up.
2. I completely forgot about the cli, since I don't use it. I'll see about wiring that up, too.

tessus commented:
> 1. Smart idea about a fixture test. I'll get that wired up.

Great, thanks.

>     2. I completely forgot about the cli, since I don't use it. I'll see about wiring that up, too.

I suggest to wait until we hear back from Orhun. I don't want you to invest time, in case he doesn't want to add this. I doubt it, but just to be safe...
I think he is quite busy these days, so it might take a while before we hear back.

anoldguy committed:
chore: address clippy lints in password protection code

Inlines format string arguments and replaces unwrap() with expect()
in tests to satisfy clippy warnings treated as errors in CI.

anoldguy commented:
Do you generally prefer a 'merge master into branch' approach, or 'rebase branch off current master and force push the PR' approach?

tessus commented:
@anoldguy any chance you can do the rebase?

tessus commented:
Not sure what to do with this one.

orhun commented:
@tessus Are you fine with adding this feature? If so, we can rebase and continue on a new branch (new PR)

tessus commented:
It is pretty isolated, so it should not be a maintenance burden. I can create a new PR, but on the other side I would like to hear back from the OP: Is he no longer interested in getting this into rustypaste or is he just busy.

anoldguy committed:
test: add fixture tests for protected file uploads

Covers the upload and duplicate upload flows end to end: parsing the
URL and generated password out of the response, confirming the content
and password sidecar land in the protected directory, and verifying
that Bearer and Basic auth gate downloads while missing or wrong
credentials get the same 404 as a missing file. Requested during
review.

anoldguy committed:
fix: allow deleting protected files through the delete endpoint

The delete handler only looked in the upload root, so protected files
living under the protected directory came back 404 and could never be
removed through the API. Writing the delete fixture surfaced this.

The handler now falls back to the protected directory the same way
serve does, and keeps the existing ordering: content file first, then
the password sidecar. A unit test and the fixture cover the full
delete flow, including the second delete returning not found.

tessus reviewed (approved):
Besides the Cargo.lock conflict, LGTM. Thanks.

tessus committed:
chore: resolve conflict

tessus committed:
Merge branch 'master' into feat/add-password-protected-files

tessus committed:
fix: generate-lockfile

tessus committed:
fix: unresolved import

orhun reviewed (approved):
thanks!
